### PR TITLE
OBAListView (replace IGListKit)

### LINKED ISSUE #290: IGListKit randomly crashing
Similar to https://github.com/OneBusAway/onebusaway-iphone/issues/1479.

This happens at randomly when IGListKit updates, likely when performing a move.
Not sure if this is iOS 14 only, but it's become more prevalent in Xcode 12.

### PR DESCRIPTION
**Pull request by ualch9**
Displays information as a vertical-scrolling list, a la TableView.

~~There are two separate `OBAListViewConfigurator`s under-the-hood. One is for iOS 14+, the other is for iOS 13.  iOS 14 uses `UICollectionLayoutListConfiguration` because its user experience is much better than our custom table view-like implementation.~~

## Todo
- [x] Rows with custom views
- [x] Cell actions (on tap, swipe, etc.)
- [x] Collapsible headers
- [x] Default rows UI (iOS 13)
- [x] Section header styling (iOS 13)

## ~~Goal~~
~~iOS 13 users get the same user experience as using IGListKit and iOS 14 users will \~automagically\~ get the upgraded experience.~~

| ~~Collection View (iOS 13) - Fade animation~~ | ~~Collection View (iOS 14) - Cell moves animation~~ |
| -------- | --------- |
 |  |  |

## How it looks right now
```swift
struct Person: OBAListViewItem {
    var name: String
    var address: String

    func listViewConfigurationForThisItem(_ listView: OBAListView) -> OBAListContentConfiguration {
        return OBAListContentConfiguration(image: UIImage(systemName: "person.fill"), text: name, secondaryText: address, appearance: .subtitle, accessoryType: .none)
    }
}

class PersonsListView: OBAListView {
    override func sectionsForList() -> [OBAListViewSection] {
        [
            OBAListViewSection(id: "1", title: "B", contents: [
                Person(name: "Bo", address: "123 Main St"),
                Person(name: "Bob", address: "456 Broadway"),
                Person(name: "Bobby", address: "asdfasdkfjad")
            ]),
            OBAListViewSection(id: "2", title: "C", contents: [
                Person(name: "Ca", address: "193 Lochmere Ln"),
                Person(name: "Cab", address: "anywhere u want"),
                Person(name: "Cabby", address: "huh")
            ])
        ]
    }
}
```

| iOS 13 | iOS 14 |
| -------- | --------- |
 |  |  |

UI work is needed for iOS 13.

### CODE REVIEW

**Comment by ualch9** [comment]
Decided to scrap the idea of using UICollectionLayoutListConfiguration for iOS 14, I'm concerned that the multi-platform implementation will be too messy to maintain.

| Collapsed sections | Sticky header |
| --- | --- |
| IMG_1686 | IMG_1687 |

**Comment by aaronbrethorst** [comment]
🎉 I'll review soon!

**Comment by ualch9** [comment]
@aaronbrethorst, made a quick naming change just now.

**Comment by ualch9** [comment]
There is a potential flaw with how I've written `createLayout()` when it comes to sections with no headers. Sections with `nil` titles will still display a header, which it shouldn't.

**Comment by aaronbrethorst** [comment]
@ualch9 I finally have time to review this! Sorry it's taken me so long! Can you rebase against `master`?

**Comment by aaronbrethorst** [comment]
I'm seeing a couple dozen SwiftLint issues. Can you install it and make sure to clean those up? (instructions are in the README)

**Comment by aaronbrethorst** [comment]
Also, you'll want to add `type_name` to the list of Swiftlint's disabled rules.

```
disabled_rules:
  - function_parameter_count
  - identifier_name
  - statement_position
  - line_length
  - todo
  - private_over_fileprivate
  - type_name

```

File: OBAKit/Alerts/AgencyAlertsViewController.swift [hunk 1]
```diff
@@ -42,36 +47,29 @@ class AgencyAlertsViewController: UIViewController,
...
-        addChildController(collectionController)
-        collectionController.view.pinToSuperview(.edges)
-        collectionController.collectionView.addSubview(refreshControl)
     }
 
     public override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
-
-        collectionController.reload(animated: false)
+        self.reloadServerData()
```
Comment (aaronbrethorst) [inline OBAKit/Alerts/AgencyAlertsViewController.swift:66]: remove `self.` here and elsewhere whenever possible.

File: OBAKit/Alerts/AgencyAlertViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+//  AgencyAlertViewModel.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/10/20.
+//
+
+import OBAKitCore
+
+extension AgencyAlert {
+    struct ListViewModel: OBAListViewItem {
```
Comment (aaronbrethorst) [inline OBAKit/Alerts/AgencyAlertViewModel.swift:11]: please add a one sentence doc comment to this.

File: OBAKit/Controls/ListView/Helpers/OBAListViewDebug.swift [hunk 3]
```diff
@@ -0,0 +1,115 @@
+//
+//  OBAListViewPreview+CustomContent.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/8/20.
+//
+
+import Foundation
+
+#if DEBUG
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Helpers/OBAListViewDebug.swift:10]: please add a one sentence comment to this file to explain what the structs and classes in it are for.

File: OBAKit/Controls/ListView/Helpers/ReuseIdentifierProviding.swift [hunk 4]
```diff
@@ -0,0 +1,17 @@
+//
+//  ReuseIdentifierProviding.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/2/20.
+//
+
+public protocol ReuseIdentifierProviding: UICollectionViewCell {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Helpers/ReuseIdentifierProviding.swift:8]: please add a one sentence doc comment to this.

File: OBAKit/Controls/ListView/OBAContentView.swift [hunk 5]
```diff
@@ -0,0 +1,28 @@
+//
+//  OBAContentView.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/8/20.
+//
+
+import Foundation
+
+/// The requirements for a content view that you create using a configuration.
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/OBAContentView.swift:10]: Can you elaborate on where an `OBAContentView`-conforming object would be used? Doesn't have to be a ton of detail, but say something like "these content views are used in our list views for blah blah blah"

File: OBAKit/Controls/ListView/Rows/OBAListRowConfiguration.swift [hunk 6]
```diff
@@ -0,0 +1,59 @@
+//
+//  OBAListRowConfiguration.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/2/20.
+//
+
+public struct OBAListRowConfiguration: OBAContentConfiguration, Hashable, Equatable {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Rows/OBAListRowConfiguration.swift:8]: please add a one sentence doc comment to this

File: OBAKit/Controls/ListView/Rows/OBAListRowViewDefault.swift [hunk 7]
```diff
@@ -0,0 +1,89 @@
+//
+//  OBAListRowViewDefault.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/4/20.
+//
+
+public class OBAListRowViewDefault: OBAListRowView {
+    static let ReuseIdentifier: String = "OBAListRowViewDefault_ReuseIdentifier"
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Rows/OBAListRowViewDefault.swift:9]: this explicit type can be removed

File: OBAKit/Controls/ListView/Rows/OBAListRowViewDefault.swift [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    override func prepareForReuse() {
+        super.prepareForReuse()
+        titleLabel.text = nil
+    }
+}
+
+// MARK: - Default ViewModel for convenience
+extension OBAListRowView {
+    public struct DefaultViewModel: OBAListViewItem {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Rows/OBAListRowViewDefault.swift:35]: Can you add a one sentence doc comment to this?

File: OBAKit/Controls/ListView/Rows/OBAListRowViewHeader.swift [hunk 9]
```diff
@@ -0,0 +1,85 @@
+//
+//  OBAListRowViewHeader.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/4/20.
+//
+
+import OBAKitCore
+
+public class OBAListRowViewHeader: OBAListRowView {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Rows/OBAListRowViewHeader.swift:10]: can you add a one sentence doc comment to this?

File: OBAKit/Controls/ListView/Rows/OBAListRowViewSubtitle.swift [hunk 10]
```diff
@@ -0,0 +1,92 @@
+//
+//  OBAListRowViewSubtitle.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/4/20.
+//
+
+public class OBAListRowViewSubtitle: OBAListRowView {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Rows/OBAListRowViewSubtitle.swift:8]: can you add a one sentence doc comment to this?

File: OBAKit/Controls/ListView/Rows/OBAListRowViewValue.swift [hunk 11]
```diff
@@ -0,0 +1,106 @@
+//
+//  OBAListRowViewValue.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/4/20.
+//
+
+public class OBAListRowViewValue: OBAListRowView {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Rows/OBAListRowViewValue.swift:8]: can you add a one sentence doc comment to this?

File: OBAKit/Controls/ListView/Rows/SupplementaryViews/OBAListViewHeaderSupplementaryView.swift [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+//
+//  Created by Alan Chu on 10/10/20.
+//
+
+// MARK: - UICollectionReusableView
+public protocol OBAListRowHeaderSupplementaryViewDelegate: class {
+    func didTap(_ headerView: OBAListRowViewHeader, section: OBAListViewSection)
+}
+
+public class OBAListRowHeaderSupplementaryView: UICollectionReusableView {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Rows/SupplementaryViews/OBAListViewHeaderSupplementaryView.swift:13]: can you add a one sentence doc comment to this?

File: OBAKit/Controls/ListView/Rows/SupplementaryViews/OBAListViewSeparatorSupplementaryView.swift [hunk 13]
```diff
@@ -0,0 +1,31 @@
+//
+//  OBAListViewSeparatorSupplementaryView.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/10/20.
+//
+
+import OBAKitCore
+
+/// A footer view displaying a separator line. Useful for faking a cell row collapse animation.
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/Rows/SupplementaryViews/OBAListViewSeparatorSupplementaryView.swift:10]: 👍

File: OBAKit/Controls/ListView/ViewModels/OBAListViewAction.swift [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// The title of the action button.
+    public var title: String?
+
+    /// The image used for the action button.
+    public var image: UIImage?
+
+    // MARK: Colors
+    /// The text color of the action button.
+    public var textColor: UIColor = .white
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/ViewModels/OBAListViewAction.swift:32]: please pull these two colors into ThemeColors

File: OBAKit/Controls/ListView/ViewModels/OBAListViewSection.swift [hunk 15]
```diff
@@ -0,0 +1,108 @@
+//
+//  OBAListViewSection.swift
+//  OBAKit
+//
+//  Created by Alan Chu on 10/3/20.
+//
+
+public struct OBAListViewSection: Hashable {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/ViewModels/OBAListViewSection.swift:8]: can you add a one sentence doc comment to this?

File: OBAKit/Controls/ListView/ViewModels/OBAListViewSection.swift [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+            } else {
+                section.boundarySupplementaryItems = [sectionHeader]
+            }
+        }
+
+        return section
+    }
+}
+
+struct OBAListViewSectionHeader: OBAListViewItem {
```
Comment (aaronbrethorst) [inline OBAKit/Controls/ListView/ViewModels/OBAListViewSection.swift:91]: can you add a one sentence doc comment to this?

**Review by aaronbrethorst — CHANGES_REQUESTED** [review]
This is looking really good! I'm seeing a crash on the Agency Alerts controller that'll need to get sorted out, plus I'd love to see a few more doc comments added while things are still fresh in your mind, but otherwise that's about it! Thanks for pulling this together!

File: OBAKit/Alerts/AgencyAlertsViewController.swift [hunk 17]
```diff
@@ -8,26 +8,31 @@
 //
 
 import UIKit
-import IGListKit
 import OBAKitCore
+import SafariServices
 
 /// Displays `AgencyAlert` objects loaded from a Protobuf feed.
 class AgencyAlertsViewController: UIViewController,
```
[the remark this answers is not in the document]
Comment (ualch9) [reply OBAKit/Alerts/AgencyAlertsViewController.swift:15]: I've randomly encountered this before, but haven't been able to reliably reproduce when debugging. I've implemented a potential fix.

**Review by aaronbrethorst — APPROVED** [review]
🎉 I think this is good to go! nice work!!

**Pull request merged**